dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/clients/common/UserInfoHelper.java

```diff
@@ -0,0 +1,130 @@
+ */
+public class UserInfoHelper {
+    private static final TraceComponent tc = Tr.register(UserInfoHelper.class, TraceConstants.TRACE_GROUP, TraceConstants.MESSAGE_BUNDLE);
+    private ConvergedClientConfig clientConfig = null;
+
+    public UserInfoHelper(ConvergedClientConfig config) {
+        this.clientConfig = config;
+    }
+
+    public boolean willRetrieveUserInfo() {
+        return (clientConfig.getUserInfoEndpointUrl() != null && clientConfig.isUserInfoEnabled() == true);
+    }
+
+    /**
+     * get userinfo from provider's UserInfo Endpoint if configured and active.
+     * If successful, update properties in the ProviderAuthenticationResult
+     *
+     * @return true if PAR was updated with userInfo
+     *
+     */
+    public boolean getUserInfo(ProviderAuthenticationResult oidcResult,
+            SSLSocketFactory sslSocketFactory, String accessToken, String subjectFromIdToken) {
+
+        if (!willRetrieveUserInfo() || accessToken == null) {
+            return false;
+        }
+        String userInfoStr = getUserInfoFromURL(clientConfig, sslSocketFactory, accessToken);
+
+        if (userInfoStr == null) {
+            return false;
+        }
+
+        if (!isUserInfoValid(userInfoStr, subjectFromIdToken)) {
+            return false;
+        }
+
+        updateAuthenticationResultPropertiesWithUserInfo(oidcResult, userInfoStr);
+        return true;
+    }
+
+    protected void updateAuthenticationResultPropertiesWithUserInfo(ProviderAuthenticationResult oidcResult, String userInfoStr) {
+        oidcResult.getCustomProperties().put(Constants.USERINFO_STR, userInfoStr);
+    }
+
+    // per oidc-connect-core-1.0 sec 5.3.2, sub claim of userinfo response must match sub claim in id token.
+    protected boolean isUserInfoValid(String userInfoStr, String subClaim) {
+        String userInfoSubClaim = getUserInfoSubClaim(userInfoStr);
+        if (userInfoSubClaim == null || subClaim == null || userInfoSubClaim.compareTo(subClaim) != 0) {
+            Tr.error(tc, "USERINFO_INVALID", new Object[] { userInfoStr, subClaim });
+            return false;
+        }
+        return true;
+    }
+
+    protected String getUserInfoSubClaim(String userInfo) {
+        JSONObject jobj = null;
+        try {
+            jobj = JSONObject.parse(userInfo);
+        } catch (Exception e) { // ffdc
+        }
+        return jobj == null ? null : (String) jobj.get("sub");
+    }
+
+    /**
+     * Obtain userInfo from an OIDC provider
+     *
+     * @return the userInfo string, or null
+     */
+    protected String getUserInfoFromURL(ConvergedClientConfig config, SSLSocketFactory sslsf, String accessToken) {
+        String url = config.getUserInfoEndpointUrl();
+        boolean hostnameVerification = config.isHostNameVerificationEnabled();
+
+        // https required by spec, use of http is not spec compliant
+        if (!url.toLowerCase().startsWith("https:") && config.isHttpsRequired()) {
```

Comment:
Is HTTPS always required by the spec? Or only if the client requires it? If the former is the case, should the `config.isHttpsRequired()` check be taken out?

Reply:
it is required by spec but we allow non-spec behavior with isHttpsRequired, so we'll warn about it.